feat: add duplicate role functionality with permission copying

Shanuka-98 opened the pull request:
**Related Issue:**

Closes #32030  

**Description:**

This PR introduces the ability to duplicate an existing Role, including all its associated permissions, directly from the Role doctype form view. This significantly speeds up the process of creating new roles based on existing ones and reduces the potential for manual configuration errors.

This addresses the feature request outlined in issue #32030. 

**Changes Implemented:**

* Added a "Duplicate" button to the Role doctype form view. * Implemented client-side logic (in `frappe/core/doctype/role/role.js`?) to:
    * Handle the "Duplicate" button click.
    * Show a prompt asking the user for the name of the new role.
    * Call the new server-side duplication method.
* Added a whitelisted server-side method (in `frappe/core/doctype/role/role.py`?) named `duplicate_role` which:
    * Takes the source Role name and the desired new Role name as arguments.
    * Performs validation (e.g., checks if the new role name already exists).
    * Creates the new Role document.
    * Copies all `Role Permission` entries from the source role to the new role.
* Redirects the user to the newly created Role's form upon successful duplication.
* Added appropriate success and error messages.

**How to Test:**

1.  Ensure you have the latest changes from this PR.
2.  Navigate to the Role list (`/app/role`).
3.  Open any existing Role (e.g., "Website Manager").
4.  Locate and click the new "Duplicate" button.
5.  In the prompt, enter a name for the new role (e.g., "Cloned Website Manager"). Click Confirm/Submit.
6.  A success message appears.
7.  A new Role named "Cloned Website Manager" now exists in the Role list.
8.  You are automatically redirected to the "Cloned Website Manager" form view.
9.  Go to Role Permissions Manager (`/app/role-permission-manager`) or inspect the "Cloned Website Manager" role document directly. Confirm that its permissions exactly match the original "Website Manager" role.


backport version-14
backport version-15

Shanuka-98 committed:
feat: add duplicate role functionality with permission copying

Shanuka-98 committed:
feat: add validation for existing role names during duplication

Shanuka-98 committed:
refactor: improve code formatting and consistency in role duplication functionality

IMS94 commented on frappe/core/doctype/role/role.py:
```diff
@@ -87,6 +88,89 @@ def update_user_type_on_change(self):
+		return target_doc
+
+	else:
+		if not new_role_name:
```
This scenario won't occur as `new_role_name` is required in the dialog.

IMS94 commented on frappe/core/doctype/role/role.py:
```diff
@@ -87,6 +88,89 @@ def update_user_type_on_change(self):
 
 
+@frappe.whitelist()
+def duplicate_role(source_name, new_role_name=None, target_role=None):
```
`target_role` is not sent from the frontend js. So, no point in having any logic around that. Please remove this param.

IMS94 commented on frappe/core/doctype/role/role.py:
```diff
@@ -87,6 +88,89 @@ def update_user_type_on_change(self):
+	        target_role: Role to copy permissions to
+	        perms: Permissions data (optional, will fetch if not provided)
+	"""
+	if not perms:
```
Don't consider the scenario where `perms` are not set. We have only 1 usage for this function, and therefore we can assume we pass that always.

Shanuka-98 committed:
feat: simplify role duplication by removing target role option and enhancing permission copying

Shanuka-98 committed:
fix: remove unnecessary comment on alert dismissal in role duplication success message

IMS94 reviewed (commented):
LGTM!

cogk commented:
Nice!

We added the feature last year in our Frappe fork but I didn't take the time to port it to Frappe :face_in_clouds:.
You might want to do the "duplication" in a **background job as it can take a long time in some cases**:

* https://gitlab.com/dokos/dodock/-/merge_requests/1693/diffs
* https://gitlab.com/dokos/dodock/-/merge_requests/5205/diffs

Shanuka-98 committed:
feat: enhance role permission copying by adding and updating permissions directly

Shanuka-98 commented:
> Nice!
> 
> We added the feature last year in our Frappe fork, but I didn't take the time to port it to Frappe 😶‍🌫️. You might want to do the "duplication" in a **background job as it can take a long time in some cases**:
> 
> * https://gitlab.com/dokos/dodock/-/merge_requests/1693/diffs
> * https://gitlab.com/dokos/dodock/-/merge_requests/5205/diffs

Hi @cogk,
Thanks for the feedback. Your code looks clean. I have referenced it and updated my code, except for the background job part
Because it's unnecessary in this feature, permission duplication won't take longer unless it has a larger number of permissions; besides that, we won't be able to show a successful copied permissions message.

Shanuka-98 committed:
fix: reorder import statements for consistency in role.py

IMS94 commented:
@sagarvora @ankush @DipenFrappe can we get this improvement added to the roles? This has come up as a requirement when allowing readonly access to ERPNext accounts for external accountants. Rather than spending time creating a completely new set of role, it's convenient to simply copy an existing role with the permissions and edit that accordingly. The author @Shanuka-98 is from my team. Appreciate if you can check this and comment your opinion on this.

iamejaaz commented on frappe/core/doctype/role/role.py:
```diff
@@ -87,6 +88,75 @@ def update_user_type_on_change(self):
 				user.save()
 
 
+@frappe.whitelist()
```
IMO, the duplication process should be handled in the background, as it may take some time. If needed, you can notify the user using real-time events.

iamejaaz commented:
@Mergifyio update

iamejaaz commented:
@Shanuka-98 @IMS94 The Duplicate button is already available, so instead of using the same name, it can be renamed to something like 'Clone' or any other suitable label.

akhilnarang commented:
@mergifyio rebase

akhilnarang commented:
> @Shanuka-98 @IMS94 The Duplicate button is already available, so instead of using the same name, it can be renamed to something like 'Clone' or any other suitable label.

@Shanuka-98 can you implement the changes mentioned by Ejaaz?

iamejaaz commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'version-15-hotfix' into feat-32030-role-duplication

clyao commented:
When will it be merged into the dev branch

iamejaaz commented:
> When will it be merged into the dev branch

This code doesn't seem to run locally. I tried running it, but it fails every time.
